### PR TITLE
Updated Rate-limiting library

### PR DESCRIPTION
**Pull request by paulo-raca**
I've been working on the `main` branch for convenience, however I'm creating this PR to get some feedback on the library.

It is mostly the same thing we had in ng-evangelion for a while, but with lots of improvements:
- Support for throttling: If your ratelimit check would have failed it can now delay the request a bit instead of returning an error immediately
- More implementations (I tried to keep implementations code as similar as possible):
  - `memory`: Fast implementation for single-server instances
  -  `redis_lua`: (Uses a lua script to make the check and update redis state atomically -- This is what we had before)
  -  `redis_module`: Uses a custom redis module to create a fast and atomic ratelimit operation
  - `redis_unlocked`: Uses 2 separate GET/SET calls to redis. This is slower and non-atomic, but good enough if you cannot run custom code in your redis server

- More tests, fixed a few corner cases
- CI, Lint, etc

### CODE REVIEW

File: rate_limiter.go [hunk 1]
```diff
@@ -4,71 +4,101 @@ package ratelimit
...
 )
 
 // A rate limit for some operation
 type Limit struct {
-	Name     string  // name added to storage keys
-	Cost     float64 // cost (in time) for 1 operation
-	Capacity float64 // time necessary to achieve max burst capacity
+	Key         string        // Identifier, used to distinguish different resources being rate-limited
+	Comment     string        // Internal comment
+	Period      time.Duration // cost (in time) for 1 operation
```
Comment (paulo-raca) [inline rate_limiter.go:24]: nit: Maybe call it `Interval` instead?

File: util/math.go
Comment (josephmgardner) [inline util/math.go:1]: Why not use the builtin `min` and `max` functions?
https://pkg.go.dev/builtin#min
https://pkg.go.dev/builtin#max

File: util/math.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	var ret T
+	for i, v := range vals {
+		if i == 0 || v > ret {
+			ret = v
+		}
+	}
+	return ret
+}
+
+func DivRoundUp[T constraints.Integer](a T, b T) T {
```
Comment (josephmgardner) [inline util/math.go:25]: Is this used anywhere? Searching for it only comes up with this result.
Also nit: the name makes it sound like it divides `a` by `b` and rounds up, which isn't quite what it does.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,147 @@
...
+
+# Rate-limiting parameters
+
+There are many different algorithms (and implementations) out there, each one using different set of parameters, which can be quite confusing.
+
+However they are usually equivalent, and it is easy to convert between them!
+
+| Name                   | Description                                                                          | Example               | Equivalence                      |
+| ---------------------- | ------------------------------------------------------------------------------------ | --------------------- | -------------------------------- |
+| `interval` or `period` | amount of time that should be elapsed between requests                               | a request every 10s   | `window/burst`, `1/rate`         |
```
Comment (josephmgardner) [inline README.md:37]:
```suggestion
| `interval` or `period` | Amount of time between requests                               | a request every 10s   | `window/burst`, `1/rate`         |
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,147 @@
...
+| `rate` or `frequency`  | How often requests can be made                                                       | 6 requests per minute | `burst/window`, `1/interval`     |
+| `burst`                | Maximum amount of requests that can be executed at once                              | 50 requests           | `window/interval`, `window*rate` |
+| `window`               | Amount of time to be considered when deciding if a request exceeds to the rate limit | 5-minute window       | `interval*burst`, `burst/rate`   |
```
Comment (josephmgardner) [inline README.md:40]:
```suggestion
| `window`               | Amount of time to consider when deciding if a request exceeds the rate limit | 5-minute window       | `interval*burst`, `burst/rate`   |
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,147 @@
...
+
+In this implementation we use `interval` and `window`. Having all parameters (and the clock) in time units is convenient and allows a simpler implementation.
+
+# Comparisson of Rate Limiting Algorithms
```
Comment (josephmgardner) [inline README.md:44]:
```suggestion
# Comparison of Rate Limiting Algorithms
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,147 @@
...
+            # new requests will only be allowed after the interval has elapsed.
+            self.ready_after = now + self.interval
+            return True
+```
+
+## Fixed Window
+
+On fixed window algorithm, time is divided in discrete intervals, like every full minute, hour or day.
+
+During that time window you can distribute the number of requests in any way you want: Evenly spread during the interval, all at once or something in-between. Either way, once exausted you will need to wait until the next window boundary (e.g., next midnight, next full hour or next full minute).
```
Comment (josephmgardner) [inline README.md:75]:
```suggestion
During that time window you can distribute the number of requests in any way you want: evenly spread during the interval, all at once, or something in-between. Either way, once exhausted you will need to wait until the next window boundary (e.g., next midnight, next full hour or next full minute).
```

File: README.md [hunk 7]
```diff
@@ -1,2 +1,147 @@
...
+
+This algorithm is very easy to reason about and simple to implement, but has its drawbacks:
```
Comment (josephmgardner) [inline README.md:77]:
```suggestion
This algorithm is very easy to reason about and simple to implement, but has a major drawback:
```

File: README.md [hunk 8]
```diff
@@ -1,2 +1,147 @@
...
+It goes immediately to fully burst-ready at the start of each window. On worst case it allows 2X the burst limit to be served in a short interval.
```
Comment (josephmgardner) [inline README.md:78]:
```suggestion
it goes immediately to fully burst-ready at the start of each window. In the worst case it allows 2x the burst limit to be served in a short interval.
```

File: README.md [hunk 9]
```diff
@@ -1,2 +1,147 @@
...
+            # update the request counter
+            self.count+=1
+            return True
+```
+
+## Sliding Window
+
+This is conceptually similar to the fixed window algorithm, but we are now counting how many requests were executed in the last `window` seconds (instead of examining a fixed interval).
+
+This has some advantages over the fixed window algorithm: It avoids the 2X burst problem near the end of the window, and it doesn't recovers the full burst capacity at predetermined times.
```
Comment (josephmgardner) [inline README.md:110]:
```suggestion
This has some advantages over the fixed window algorithm: it avoids the 2x burst problem near the end of the window, and it doesn't recover the full burst capacity at predetermined times.
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,147 @@
...
+
+However the implementation is considerably more complex and requires more memory:
+
+- An exact solution now requires keeping track of all request timestamps inside the window
+- An efficient (but approximate) solution can be achieved by adding the counter of several smaller fixed windows (e.g., you can simulate the counter of a 1-minute sliding window by adding the last 60 1-second fixed windows - [Take a look at this Figma blog post for more info](https://www.figma.com/blog/an-alternative-approach-to-rate-limiting/#fixed-window-counters)
```
Comment (josephmgardner) [inline README.md:115]:
```suggestion
- An efficient (but approximate) solution can be achieved by adding the counter of several smaller fixed windows (e.g., you can simulate the counter of a 1-minute sliding window by adding the last 60 1-second fixed windows - [take a look at this Figma blog post for more info](https://www.figma.com/blog/an-alternative-approach-to-rate-limiting/#fixed-window-counters).
```

File: README.md [hunk 11]
```diff
@@ -1,2 +1,147 @@
...
+
+(TODO: Improve this) Another quirk of sliding windows is that capacity increases in the same pattern as the usage in the previous window. E.g., If you get burst of 10 requests, an idle interval and then another burst that exausts your rate limit, in the next window you will also get an increase of 10 in the burst capacity, then no changes during an interval and then full burst capacity. Instead it is preferable to increase it continuously every `window/burst` seconds (See Leaky Bucket below)
```
Comment (josephmgardner) [inline README.md:117]: Still TODO?

Comment (josephmgardner) [inline README.md:117]:
```suggestion
(TODO: Improve this) Another quirk of sliding windows is that capacity increases in the same pattern as the usage in the previous window. E.g., If you get burst of 10 requests, an idle interval and then another burst that exausts your rate limit, in the next window you will also get an increase of 10 in the burst capacity, then no changes during an interval and then full burst capacity. Instead it is preferable to increase it continuously every `window/burst` seconds (see Leaky Bucket below).
```

File: README.md [hunk 12]
```diff
@@ -1,2 +1,147 @@
...
+            # insert the current timestamp in the window
+            self.request_timestamps.append(now)
+            return True
+```
+
+## GCRA (aka Leaky Bucket)
+
+The algorithm we use, is awesome.
+
+TODO
```
Comment (josephmgardner) [inline README.md:147]: I appreciate the rundown of the other algorithms but this is probably the most important one to explain, since it's the one the library actually uses - is the plan to add this before merge @paulo-raca?

File: rate_limiter.go [hunk 13]
```diff
@@ -79,42 +109,58 @@ func (cl *CompositeLimit) Flatten(keys ...interface{}) ([]*Limit, error) {
...
+	Remaining:    math.MaxInt64,
+	ReadyAfter:   0,
+	ResetAfter:   0,
 }
 
 func (r *RateLimitResult) AsError() error {
 	if !r.Allowed {
 		// TODO: Maybe add Retry-After and X-RateLimit-* headers?
-		return fmt.Errorf("Too Many Requests")
+		return fmt.Errorf("too Many Requests")
```
Comment (josephmgardner) [inline rate_limiter.go:158]:
```suggestion
		return fmt.Errorf("too many requests")
```

File: redis_lua.lua [hunk 14]
```diff
@@ -11,57 +9,73 @@ if now == nil then  -- Use timestamp from Redis
 	-- which is good enough for up rate limit of up to the MHz range
 	local epoch = 1672531200  -- 2023-01-01
 	now = redis.call("TIME")
-	now = (now[1] - epoch) + now[2] * 1e-6
+	now = (now[1] - epoch) * 1e9 + now[2] * 1e3
```
Comment (josephmgardner) [inline redis_lua.lua:12]: What changed here?

**Review by josephmgardner — APPROVED** [review]

File: util/math.go
Comment (paulo-raca) [reply util/math.go:1]: I really looked for them! :scream: 

All I had found was `math.Min()` / `math.Max()`, but those are only for float64

File: redis_lua.lua (hunk 14 above)
Comment (paulo-raca) [reply redis_lua.lua:12]: The counter was the number of seconds as a float value and is now the number of nanos
(which is still a float value in the lua version, but an int64 everywhere else)
